sparse/coo.py

```diff
@@ -2034,113 +1988,127 @@ def _grouped_reduce(x, groups, method, **kwargs):
+        return func(**kwargs)
 
-    # Start with an empty list. This may reduce computation in many cases.
-    data_list = []
-    coords_list = []
+    if len(args) == 1:
+        return _elemwise_unary(func, args[0], **kwargs)
 
-    # Add the matched part.
-    matched_coords = _get_matching_coords(self_coords[:, matched_self],
-                                          other_coords[:, matched_other],
-                                          self_shape, other_shape)
+    return _elemwise_n_ary(func, *args, **kwargs)
 
-    data_list.append(func(self_data[matched_self],
-                          other_data[matched_other],
-                          *args, **kwargs))
-    coords_list.append(matched_coords)
 
-    self_func = func(self_data, other_zero, *args, **kwargs)
-    # Add unmatched parts as necessary.
-    if (self_func != func_zero).any():
-        self_unmatched_coords, self_unmatched_func = \
-            _get_unmatched_coords_data(self_coords, self_func, self_shape,
-                                       result_shape, matched_self,
-                                       matched_coords)
+def _elemwise_n_ary(func, *args, **kwargs):
+    """
+    Apply a function to any number of arguments with broadcasting.
 
-        data_list.extend(self_unmatched_func)
-        coords_list.extend(self_unmatched_coords)
+    Parameters
+    ----------
+    func : Callable
+        The function to apply to arguments. Must support broadcasting.
+    args : list
+        Input :obj:`COO` or :obj:`numpy.ndarray`s.
+    kwargs : dict
+        Additional arguments to pass to the function.
 
-    other_func = func(self_zero, other_data, *args, **kwargs)
+    Returns
+    -------
+    COO
+        The output array.
+
+    Raises
+    ------
+    ValueError
+        If the input shapes aren't compatible or the result will be dense.
+    """
+    args = list(args)
+
+    for arg in args:
+        if isinstance(arg, COO):
+            arg.sum_duplicates()
+
+    args_zeros = tuple(_zero_of_dtype(arg.dtype)[()] for arg in args)
+
+    func_value = func(*args_zeros, **kwargs)
+    func_zero = _zero_of_dtype(np.dtype(func_value))
+    if func_value != func_zero:
+        raise ValueError("Performing this operation would produce "
+                         "a dense result: %s" % str(func))
+    data_list = []
+    coords_list = []
 
-    if (other_func != func_zero).any():
-        other_unmatched_coords, other_unmatched_func = \
-            _get_unmatched_coords_data(other_coords, other_func, other_shape,
-                                       result_shape, matched_other,
-                                       matched_coords)
+    for mask in product([True, False], repeat=len(args)):
+        if not any(mask):
+            continue
 
-        coords_list.extend(other_unmatched_coords)
-        data_list.extend(other_unmatched_func)
+        ci, di = _unmatch_coo(func, args, mask, **kwargs)
+
+        coords_list.extend(ci)
+        data_list.extend(di)
```

Comment:
This confuses me and seems concerning.  I see that this was a main point of your conversation with @shoyer earlier.  I probably have some thinking to do on this problem before I'm able to reasonably comment on this.